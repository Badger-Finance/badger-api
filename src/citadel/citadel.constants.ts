import { Protocol } from '@badger-dao/sdk';

export const CITADEL_SUBGRAPH_URL = 'https://api.thegraph.com/subgraphs/name/axejintao/citadog';
export const CITADEL_KNIGHTS_ROUND_SUBGRAPH_URL =
  'https://api.thegraph.com/subgraphs/name/namelsking/citadel-knighting-rounds';

export const CITADEL_KNIGHTS = [
  Protocol.Terra,
  Protocol.Frax,
  Protocol.Alchemix,
  Protocol.Convex,
  Protocol.Redacted,
  Protocol.JonesDao,
  Protocol.Tokemak,
  Protocol.Tribe,
  Protocol.Ren,
];
