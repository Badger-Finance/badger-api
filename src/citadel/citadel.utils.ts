import { ethers } from 'ethers';
import { getDataMapper } from '../aws/dynamodb.utils';
import { KeyedDataBlob } from '../aws/models/keyed-data-blob.model';
import { CitadelData, CTIADEL_DATA } from './destructors/citadel-data.destructor';
import { Nullable } from '../utils/types.utils';
import { CitadelRewardsSnapshot } from '../aws/models/citadel-rewards-snapshot';
import BadgerSDK from '@badger-dao/sdk';
import { RewardEventType, RewardEventTypeEnum } from '@badger-dao/sdk/lib/citadel/enums/reward-event-type.enum';
import { ListRewardsEvent } from '@badger-dao/sdk/lib/citadel/interfaces/list-rewards-event.interface';
import { CitadelRewardType } from '@badger-dao/sdk/lib/api/enums/citadel-reward-type.enum';
import { CitadelRewardsAprBlob } from './interfaces/citadel-rewards-apr-blob.interface';

export async function queryCitadelData(): Promise<CitadelData> {
  const mapper = getDataMapper();

  let blob;
  for await (const item of mapper.query(KeyedDataBlob, { id: CTIADEL_DATA }, { limit: 1 })) {
    blob = item;
  }
  if (!blob) {
    throw new Error('No stored citadel data!');
  }

  return new CitadelData(blob.data);
}

export async function getRewardsOnchain(
  sdk: BadgerSDK,
  type: RewardEventType,
  fromLastBlock = false,
): Promise<ListRewardsEvent[]> {
  let chainRewards: ListRewardsEvent[] = [];

  let lastRewardBlock: Nullable<number>;

  if (fromLastBlock) lastRewardBlock = Number((await getLastRewardByType(type))?.block) + 1;

  try {
    const getOpts: { filter: RewardEventType; startBlock?: number } = { filter: type };
    if (lastRewardBlock) getOpts.startBlock = lastRewardBlock;

    chainRewards = await sdk.citadel.listRewards(getOpts);
  } catch (err) {
    throw new Error(`Failed to get citadel listRewards ${err}`);
  }

  return chainRewards;
}

export async function getLastRewardByType(type: string): Promise<Nullable<CitadelRewardsSnapshot>> {
  const mapper = getDataMapper();

  const query = mapper.query(
    CitadelRewardsSnapshot,
    { payType: type },
    { limit: 1, scanIndexForward: false, indexName: 'IndexCitadelRewardsDataPayTypeBlock' },
  );

  let lastReward = null;

  try {
    for await (const reward of query) {
      lastReward = reward;
    }
  } catch (e) {
    console.error(`Failed to get citadel reward from ddb for type: ${type}; ${e}`);
  }

  return lastReward;
}

export async function getRewardsAprForDataBlob(): Promise<CitadelRewardsAprBlob> {
  const mapper = getDataMapper();

  const summaryCountTpl = {
    apr: 0,
    count: 0,
  };
  const summary = {
    overall: { ...summaryCountTpl },
    [CitadelRewardType.Citadel]: { ...summaryCountTpl },
    [CitadelRewardType.Funding]: { ...summaryCountTpl },
    [CitadelRewardType.Tokens]: { ...summaryCountTpl },
    [CitadelRewardType.Yield]: { ...summaryCountTpl },
  };

  const query = mapper.query(
    CitadelRewardsSnapshot,
    { payType: RewardEventTypeEnum.ADDED },
    {
      indexName: 'IndexCitadelRewardsDataPayType',
      filter: {
        type: 'GreaterThanOrEqualTo',
        object: Date.now(),
        subject: 'finishTime',
      },
    },
  );

  try {
    for await (const reward of query) {
      summary.overall.apr += reward.apr || 0;
      summary.overall.count += 1;

      const rewardTypeKey = getRewardsEventTypeMapped(<string>reward.dataType);

      summary[rewardTypeKey].apr += reward.apr || 0;
      summary[rewardTypeKey].count += 1;
    }
  } catch (e) {
    console.error(`Failed to get citadel reward from ddb`);
  }

  return Object.keys(summary).reduce((acc, val) => {
    const assertKey = <keyof CitadelRewardsAprBlob>val;
    acc[assertKey] = summary[assertKey].apr / summary[assertKey].count;
    return acc;
  }, <CitadelRewardsAprBlob>{});
}

export function dataTypeRawKeyToKeccak(key: string): string {
  let hashedKey: string;

  try {
    hashedKey = ethers.utils.keccak256(ethers.utils.formatBytes32String(key));
  } catch (e) {
    console.warn(`Failed to hash raw key: ${key}. ${e}`);
    hashedKey = '';
  }

  return hashedKey;
}

export function getRewardsEventTypeMapped(hashKey: string): CitadelRewardType {
  return (
    {
      // Note! This is still not actual keys, need to figure out, what value
      // is hashed in the event
      [dataTypeRawKeyToKeccak('Citadel')]: CitadelRewardType.Citadel,
      [dataTypeRawKeyToKeccak('Funding')]: CitadelRewardType.Funding,
      [dataTypeRawKeyToKeccak('Yield')]: CitadelRewardType.Yield,
      [dataTypeRawKeyToKeccak('Tokens')]: CitadelRewardType.Tokens,
    }[hashKey] || CitadelRewardType.Tokens
  );
}
