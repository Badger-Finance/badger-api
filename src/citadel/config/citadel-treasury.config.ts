import { Network } from '@badger-dao/sdk';
import { SUPPORTED_CHAINS } from '../../chains/chain';

const citadelChains = new Set([Network.Ethereum]);

const chains = SUPPORTED_CHAINS.filter((c) => citadelChains.has(c.network));
const tokens = new Set(chains.flatMap((c) => Object.keys(c.tokens)));
const vaults = new Set(chains.flatMap((c) => c.vaults.map((v) => v.vaultToken)));
vaults.forEach((v) => tokens.delete(v));

export const TRACKED_TOKENS = [...tokens];
export const TRACKED_VAULTS = [...vaults];

export const CITADEL_TREASURY_ADDRESS = '0x38724146C8dc1Aa49c3395091cf86B789c37F52c';
