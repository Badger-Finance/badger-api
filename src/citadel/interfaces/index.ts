export { CitadelTreasurySummary } from './citadel-treasury-summary.interface';
