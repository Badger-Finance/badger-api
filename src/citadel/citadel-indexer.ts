import { Erc20__factory, formatBalance, Protocol } from '@badger-dao/sdk';
import { getCachedAccount } from '../accounts/accounts.utils';
import { Ethereum } from '../chains/config/eth.config';
import { getPrice } from '../prices/prices.utils';
import { VaultsService } from '../vaults/vaults.service';
import { getVaultDefinition } from '../vaults/vaults.utils';
import { CITADEL_TREASURY_ADDRESS, TRACKED_TOKENS, TRACKED_VAULTS } from './config/citadel-treasury.config';
import { TreasuryPosition } from '../treasury/interfaces/treasy-position.interface';
import { TreasurySummary } from '../treasury/interfaces/treasury-summary.interface';
import { getDataMapper } from '../aws/dynamodb.utils';
import { TreasurySummarySnapshot } from '../aws/models/treasury-summary-snapshot.model';
import { TOKEN_PRICE_DATA, TREASURY_HISTORIC_DATA, TREASURY_SNAPSHOT_DATA } from '../config/constants';

export async function snapshotTreasury() {
  const chain = new Ethereum();
  const sdk = await chain.getSdk();

  const lookupTokens = TRACKED_TOKENS.concat(TRACKED_VAULTS);
  const tokenInformation = await sdk.tokens.loadTokens(lookupTokens);

  const vaultPositions = new Set(TRACKED_VAULTS);
  const treasuryPositions: TreasuryPosition[] = await Promise.all(
    Object.entries(tokenInformation).map(async (e) => {
      const [address, token] = e;
      const isFarming = vaultPositions.has(address);

      const contract = Erc20__factory.connect(address, sdk.provider);
      const treasuryBalance = await contract.balanceOf(CITADEL_TREASURY_ADDRESS);

      const { price } = await getPrice(address);
      const balance = formatBalance(treasuryBalance, token.decimals);
      const value = price * balance;

      if (!isFarming) {
        return {
          ...token,
          balance,
          value,
          apr: 0,
          protocol: 'none',
        };
      }

      const { apr, minApr, maxApr } = await VaultsService.loadVault(chain, getVaultDefinition(chain, address));
      const { boost } = await getCachedAccount(chain, CITADEL_TREASURY_ADDRESS);
      const percentile = boost / 2000;

      let treasuryApr;
      if (minApr && maxApr) {
        treasuryApr = minApr + (maxApr - minApr) * percentile;
      } else {
        treasuryApr = apr;
      }

      return {
        ...token,
        balance,
        value,
        apr: treasuryApr,
        protocol: Protocol.Badger,
      };
    }),
  );

  const positions = treasuryPositions.filter((p) => p.value > 0);
  const value = positions.reduce((total, position) => (total += position.value), 0);

  let tvlApr = 0;
  positions.forEach((p) => {
    if (p.apr > 0) {
      tvlApr += p.apr * p.value;
    }
  });

  const treasuryYield = value === 0 ? 0 : tvlApr / value;

  const treasurySummary: TreasurySummary = {
    address: CITADEL_TREASURY_ADDRESS,
    value,
    yield: treasuryYield,
    positions,
  };

  const mapper = getDataMapper();

  console.log({ TREASURY_SNAPSHOT_DATA, TREASURY_HISTORIC_DATA, TOKEN_PRICE_DATA });
  // capture a current view of the treasury
  try {
    const currentSnapshot = Object.assign(new TreasurySummarySnapshot(), treasurySummary);
    console.log(currentSnapshot);
    await mapper.put(currentSnapshot);
  } catch (err) {
    console.error({ message: 'Unable to save Ctiadel treasury snapshot', err });
  }
}
