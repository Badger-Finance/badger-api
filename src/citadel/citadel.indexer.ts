import { formatBalance } from '@badger-dao/sdk';
import { Ethereum } from '../chains/config/eth.config';
import { TRACKED_TOKENS, TRACKED_VAULTS } from './config/citadel-treasury.config';
import { TreasuryPosition } from './interfaces/treasy-position.interface';

export async function snapshotTreasury() {
  const chain = new Ethereum();
  const sdk = await chain.getSdk();

  const lookupTokens = TRACKED_TOKENS.concat(TRACKED_VAULTS);
  const tokenBalances = await sdk.tokens.loadBalances(lookupTokens);
  const tokenInformation = await sdk.tokens.loadTokens(lookupTokens);

  const vaultPositions = new Set(TRACKED_VAULTS);
  const treasuryPositions: TreasuryPosition[] = Object.entries(tokenBalances).map((e) => {
    const [token, balance] = e;
    const isFarming = vaultPositions.has(token);
    const tokenInfo = tokenInformation[token];
    if (!isFarming) {
      return {
        token: tokenInfo,
        amount: formatBalance(balance, tokenInfo.decimals),
        value: 0,
        apr: 0,
      };
    }
    return {
      token: tokenInfo,
      amount: formatBalance(balance, tokenInfo.decimals),
      value: 0,
      apr: 0,
    };
  });
}
