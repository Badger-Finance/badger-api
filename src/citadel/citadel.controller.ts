import { CitadelTreasurySummary } from '@badger-dao/sdk/lib/api/interfaces/citadel-treasury-summary.interface';
import { Controller, Get, Inject, QueryParams, UseCache } from '@tsed/common';
import { ContentType, Summary, Returns, Description } from '@tsed/schema';
import { HistoricTreasurySummarySnapshot } from '../aws/models/historic-treasury-summary-snapshot.model';
import { ChartTimeFrame } from '../charts/enums/chart-timeframe.enum';
import { TreasuryService } from '../treasury/treasury.service';
import { CitadelService } from './citadel.service';
import { CITADEL_TREASURY_ADDRESS } from './config/citadel-treasury.config';
import { CitadelTreasurySummaryModel } from './interfaces/citadel-treasury-summary-model.interface';
import { RewardFilter } from '@badger-dao/sdk/lib/citadel/enums/reward-filter.enum';
import { CitadelRewardEventModel } from './interfaces/citadel-reward-event-model.interface';
import { CitadelRewardEvent } from './interfaces/citadel-reward-event.interface';

@Controller('/')
export class CitadelController {
  @Inject()
  citadelService!: CitadelService;

  @Inject()
  treasuryService!: TreasuryService;

  @UseCache()
  @Get('/treasury')
  @ContentType('json')
  @Summary('Citadel DAO Treasury Allocation and Metrics')
  @Description('Returns specific treasury metrics, positions, and yield')
  @Returns(200, CitadelTreasurySummaryModel)
  async loadCitadelTreasury(): Promise<CitadelTreasurySummary> {
    return this.citadelService.loadTreasurySummary();
  }

  @Get('/rewards')
  @ContentType('json')
  @Summary('Added/paid rewards list')
  @Description('List of reward events')
  @Returns(200, Array).Of(CitadelRewardEventModel)
  @Returns(400).Description('Token should be specified')
  @Returns(400).Description('User or token param is missing')
  async loadRewardsList(
    @QueryParams('token') token: string,
    @QueryParams('user') user?: string,
    @QueryParams('filter') filter?: RewardFilter,
  ): Promise<CitadelRewardEvent[]> {
    return this.citadelService.getListRewards(token, user, filter);
  }

  @UseCache()
  @Get('/history')
  @ContentType('json')
  async loadCitadelTreasuryCharts(
    @QueryParams('timeframe') timeframe = ChartTimeFrame.Day,
  ): Promise<HistoricTreasurySummarySnapshot[]> {
    return this.treasuryService.loadTreasuryChartData(CITADEL_TREASURY_ADDRESS, timeframe);
  }
}
