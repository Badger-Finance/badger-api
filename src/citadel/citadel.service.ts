import { Erc20__factory, formatBalance } from '@badger-dao/sdk';
import { CitadelTreasurySummary } from '@badger-dao/sdk/lib/api/interfaces/citadel-treasury-summary.interface';
import { Service } from '@tsed/di';
import { TOKENS } from '../config/tokens.config';
import { getPrice } from '../prices/prices.utils';
import { queryTreasurySummary } from '../treasury/treasury.utils';
import { queryCitadelData } from './citadel.utils';
import { CITADEL_TREASURY_ADDRESS } from './config/citadel-treasury.config';
import { Chain } from '../chains/config/chain.config';
import { RewardFilter } from '@badger-dao/sdk/lib/citadel/enums/reward-filter.enum';
import { BadRequest } from '@tsed/exceptions';
import { ListRewardsEvent } from '@badger-dao/sdk/lib/citadel/interfaces/list-rewards-event.interface';
import { CitadelRewardEvent } from './interfaces/citadel-reward-event.interface';

@Service()
export class CitadelService {
  async loadTreasurySummary(): Promise<CitadelTreasurySummary> {
    const [baseTreasurySummary, citadelData] = await Promise.all([
      queryTreasurySummary(CITADEL_TREASURY_ADDRESS),
      queryCitadelData(),
    ]);

    const { price } = await getPrice(TOKENS.WBTC);
    const {
      marketCapToTreasuryRatio,
      valuePaid,
      fundingBps,
      lockingBps,
      stakingBps,
      marketCap,
      supply,
      staked,
      stakedPercent,
    } = citadelData;

    const valueBtc = baseTreasurySummary.value / price;
    const valuePaidBtc = valuePaid / price;

    return {
      ...baseTreasurySummary,
      valueBtc,
      valuePaid,
      valuePaidBtc,
      marketCapToTreasuryRatio,
      fundingBps,
      stakingBps,
      lockingBps,
      marketCap,
      supply,
      staked,
      stakedPercent,
    };
  }

  async getListRewards(token: string, user?: string, filter?: RewardFilter): Promise<CitadelRewardEvent[]> {
    const chain = Chain.getChain();
    const sdk = await chain.getSdk();

    let chainRewards: ListRewardsEvent[] = [];

    try {
      chainRewards = await sdk.citadel.listRewards({ user, token, filter });
    } catch (err) {
      throw new BadRequest(`${err}`);
    }

    return Promise.all(
      chainRewards.map(async (event) => {
        const rewardEventResp: CitadelRewardEvent = {
          token: event.token,
          block: event.block,
          amount: 0,
          type: filter ?? RewardFilter.ADDED,
        };

        const tokenContract = Erc20__factory.connect(event.token, sdk.provider);
        const tokenDecimals = await tokenContract.decimals();

        rewardEventResp.amount = formatBalance(event.reward, tokenDecimals);

        if (event.user) rewardEventResp.user = event.user;

        return rewardEventResp;
      }),
    );
  }
}
