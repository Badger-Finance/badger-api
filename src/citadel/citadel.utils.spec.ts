/* eslint-disable @typescript-eslint/ban-ts-comment */
import BadgerSDK, { ListRewardsOptions, VaultV15__factory } from '@badger-dao/sdk';
import { VaultState } from '@badger-dao/sdk/lib/api';
import { CitadelRewardType } from '@badger-dao/sdk/lib/api/enums/citadel-reward-type.enum';
import { CitadelService } from '@badger-dao/sdk/lib/citadel';
import { RewardEventTypeEnum } from '@badger-dao/sdk/lib/citadel/enums/reward-event-type.enum';
import { ListRewardsEvent } from '@badger-dao/sdk/lib/citadel/interfaces/list-rewards-event.interface';
import { VaultV15 } from '@badger-dao/sdk/lib/contracts/VaultV15';
import { VaultVersion } from '@badger-dao/sdk/lib/vaults';
import { LoadVaultOptions } from '@badger-dao/sdk/lib/vaults/interfaces';
import { VaultsService } from '@badger-dao/sdk/lib/vaults/vaults.service';
import { BigNumber, ethers } from 'ethers';
import { mock } from 'jest-mock-extended';

import { CitadelRewardsSnapshot } from '../aws/models/citadel-rewards-snapshot';
import { Chain } from '../chains/config/chain.config';
import { Ethereum } from '../chains/config/eth.config';
import { TOKENS } from '../config/tokens.config';
import * as citadelGraph from '../graphql/generated/citadel';
import * as priceUtils from '../prices/prices.utils';
import { mockBadgerSdk, setupMapper, TEST_ADDR } from '../test/tests.utils';
import { fullTokenMockMap } from '../tokens/mocks/full-token.mock';
import { Nullable } from '../utils/types.utils';
import * as citadelUtils from './citadel.utils';
import {
  dataTypeRawKeyToKeccak,
  getRewardsAprForDataBlob,
  getRewardsEventTypeMapped,
  getRewardsOnchain,
  getStakedCitadelEarnings,
  getStakedCitadelPrice,
  getTokensPaidSummary,
} from './citadel.utils';
import { citadelListRewardsSdkMock } from './mocks/citadel-list-rewards.sdk.mock';
import { RewardsSnapshotModelMock } from './mocks/rewards-snapshot.model.mock';

const CURRENT_BLOCK = RewardsSnapshotModelMock[0].block;

describe('getRewardsOnchain', () => {
  let sdk: BadgerSDK;

  let listRewardsMock: jest.SpyInstance<Promise<ListRewardsEvent[]>, [options?: ListRewardsOptions | undefined]>;
  let getLastRewardByTypeMock: jest.SpyInstance<Promise<Nullable<CitadelRewardsSnapshot>>, [type: string]>;

  beforeEach(async () => {
    listRewardsMock = jest
      .spyOn(CitadelService.prototype, 'listRewards')
      .mockImplementation(async () => citadelListRewardsSdkMock);

    getLastRewardByTypeMock = jest
      .spyOn(citadelUtils, 'getLastRewardByType')
      .mockImplementation(async () => RewardsSnapshotModelMock[0]);

    sdk = await mockBadgerSdk({ currBlock: CURRENT_BLOCK });

    await sdk.ready();
  });

  it('returns rewards from the last block param', async (done) => {
    const rewards = await getRewardsOnchain(sdk, RewardEventTypeEnum.ADDED, true);

    expect(getLastRewardByTypeMock.mock.calls.length).toBe(1);
    expect(listRewardsMock).toHaveBeenCalledWith({ filter: RewardEventTypeEnum.ADDED, startBlock: CURRENT_BLOCK + 1 });
    expect(rewards).toMatchObject(citadelListRewardsSdkMock);

    done();
  });

  it('returns rewards all reward from when contract was deployed', async (done) => {
    const rewards = await getRewardsOnchain(sdk, RewardEventTypeEnum.ADDED);

    expect(getLastRewardByTypeMock.mock.calls.length).toBe(0);
    expect(listRewardsMock).toHaveBeenCalledWith({ filter: RewardEventTypeEnum.ADDED });
    expect(rewards).toMatchObject(citadelListRewardsSdkMock);

    done();
  });

  it('throws error if something went wrong in onchain call', async (done) => {
    jest.spyOn(CitadelService.prototype, 'listRewards').mockImplementation(async () => {
      throw Error('Bad things');
    });
    await expect(async () => getRewardsOnchain(sdk, RewardEventTypeEnum.ADDED)).rejects.toThrowError();

    done();
  });
});

describe('getRewardsAprForDataBlob', () => {
  it('returns rewards apr', async (done) => {
    setupMapper(RewardsSnapshotModelMock.filter((rw) => rw.payType === RewardEventTypeEnum.ADDED));

    const rwSummary = await getRewardsAprForDataBlob();

    expect(rwSummary).toMatchSnapshot();
    done();
  });
});

describe('getTokensPaidSummary', () => {
  it('should return rewards tokens paid summary', async (done) => {
    setupMapper(RewardsSnapshotModelMock.filter((rw) => rw.payType === RewardEventTypeEnum.PAID));

    const tokensPaidRwSummary = await getTokensPaidSummary();

    expect(tokensPaidRwSummary).toMatchSnapshot();
    done();
  });
});

describe('dataTypeRawKeyToKeccak', () => {
  const CITADEL_RW_TYPE_STR = 'xcitadel-locker-emissions';

  beforeEach(() => {
    jest.spyOn(console, 'warn').mockImplementation(() => void 0);
  });

  it('returns valid Keccak256 hash', () => {
    const citadelRwTypeHashed = ethers.utils.solidityKeccak256(['string'], [CITADEL_RW_TYPE_STR]);

    expect(dataTypeRawKeyToKeccak(CITADEL_RW_TYPE_STR)).toBe(citadelRwTypeHashed);
  });

  it('returns empty string for invalid type', () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    expect(dataTypeRawKeyToKeccak(null)).toBe('');
  });
});

describe('getRewardsEventTypeMapped', () => {
  const CITADEL_RW_TYPE_HASH = ethers.utils.solidityKeccak256(['string'], ['xcitadel-locker-emissions']);

  it('returns corret CitadelRewardType enum value', () => {
    expect(getRewardsEventTypeMapped(CITADEL_RW_TYPE_HASH)).toBe(CitadelRewardType.Citadel);
  });

  it('returns default enum value for unknown keys', () => {
    expect(getRewardsEventTypeMapped('unknown')).toBe(CitadelRewardType.Tokens);
  });
});

describe('getStakedCitadelPrice', () => {
  it('returns staked price and addr', async (done) => {
    jest.spyOn(VaultsService.prototype, 'loadVault').mockImplementation(async ({ address }: LoadVaultOptions) => ({
      address,
      name: 'Citadel',
      available: 500000,
      balance: 300,
      totalSupply: 1800000000,
      pricePerFullShare: 15,
      token: fullTokenMockMap[TOKENS.BADGER],
      state: VaultState.Open,
      version: VaultVersion.v1_5,
      symbol: 'BDGR',
      decimals: 18,
    }));

    jest.spyOn(priceUtils, 'getPrice').mockImplementation(async (token: string) => {
      return {
        [TOKENS.CTDL]: { address: TOKENS.CTDL, price: 21 },
      }[token];
    });

    const ethChain = mock<Ethereum>();
    ethChain.getSdk.calledWith().mockImplementation(() => mockBadgerSdk({}));

    const stakedPrice = await getStakedCitadelPrice(ethChain, fullTokenMockMap[TOKENS.BADGER]);

    expect(stakedPrice).toMatchSnapshot();
    done();
  });
});

describe('getStakedCitadelEarnings', () => {
  beforeEach(async () => {
    const citadelGraphSdkMock = {
      VaultBalance: async () => ({
        vaultBalance: {
          netShareDeposit: BigNumber.from(`${123e18}`),
          grossDeposit: BigNumber.from(`${34e18}`),
          grossWithdraw: BigNumber.from(`${566e18}`),
        },
      }),
    };
    // @ts-ignore
    jest.spyOn(citadelGraph, 'getSdk').mockImplementation(() => citadelGraphSdkMock);

    const ethChain = mock<Ethereum>();
    ethChain.getSdk.calledWith().mockImplementation(() => mockBadgerSdk({}));
    jest.spyOn(Chain, 'getChain').mockImplementation(() => ethChain);

    const vaultV15 = mock<VaultV15>();
    jest.spyOn(VaultV15__factory, 'connect').mockImplementation(() => vaultV15);
    vaultV15.getPricePerFullShare.calledWith().mockImplementation(async () => BigNumber.from(150));
  });

  it('returns staked earnings', async () => {
    const earnings = await getStakedCitadelEarnings(TEST_ADDR);
    expect(earnings).toBe(532);
  });

  it('returns 0, when cant get vaultBalance', async () => {
    const citadelGraphSdkMock = {
      VaultBalance: async () => ({ vaultBalance: null }),
    };
    // @ts-ignore
    jest.spyOn(citadelGraph, 'getSdk').mockImplementation(() => citadelGraphSdkMock);

    const earnings = await getStakedCitadelEarnings(TEST_ADDR);
    expect(earnings).toBe(0);
  });
});
