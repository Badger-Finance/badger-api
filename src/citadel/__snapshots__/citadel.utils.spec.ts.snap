// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRewardsAprForDataBlob returns rewards apr 1`] = `
Object {
  "0": 0,
  "1": 6.692294226897413,
  "2": 8.948838125220531,
  "3": 0,
  "overall": 8.046220565891284,
}
`;

exports[`getStakedCitadelPrice returns staked price and addr 1`] = `
Object {
  "address": "0x3472A5A71965499acd81997a54BBA8D852C6E53d",
  "price": 315,
}
`;

exports[`getTokensPaidSummary should return rewards tokens paid summary 1`] = `
Object {
  "0xa0FFfb6b575045f215432b3158Ffd0A9ee0454B9": 19243.263336764325,
}
`;
